### PR TITLE
Issue #1014 - added between function

### PR DESCRIPTION
**Pull request by JohnTimm**
Signed-off-by: John T.E. Timm <johntimm@us.ibm.com>

### CODE REVIEW

File: fhir-path/src/main/java/com/ibm/fhir/path/FHIRPathDateTimeValue.java [hunk 1]
```diff
@@ -60,73 +60,78 @@
...
-     * 
+     *
      * @param text
      *     the {@link String} value that is parsed into a {@link TemporalAccessor} date/time
      * @return
      *     a new FHIRPathDateTimeValue instance
      */
     public static FHIRPathDateTimeValue dateTimeValue(String text) {
-        return FHIRPathDateTimeValue.builder(DATE_TIME_PARSER_FORMATTER.parseBest(text, ZonedDateTime::from, LocalDateTime::from, LocalDate::from, YearMonth::from, Year::from), getTimePrecision(text)).build();
+//      return FHIRPathDateTimeValue.builder(PARSER_FORMATTER.parseBest(text, ZonedDateTime::from, LocalDateTime::from, LocalDate::from, YearMonth::from, Year::from), getTimePrecision(text)).build();
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/FHIRPathDateTimeValue.java:124]: can we remove it?

File: fhir-path/src/main/java/com/ibm/fhir/path/FHIRPathTimeValue.java [hunk 2]
```diff
@@ -26,9 +26,9 @@
 /**
  * A {@link FHIRPathTemporalValue} node that wraps a {@link LocalTime} value
  */
-public class FHIRPathTimeValue extends FHIRPathAbstractNode implements FHIRPathTemporalValue {    
-    private static final DateTimeFormatter TIME_PARSER_FORMATTER = new DateTimeFormatterBuilder()
-            .appendLiteral("T")
+public class FHIRPathTimeValue extends FHIRPathAbstractNode implements FHIRPathTemporalValue {
+    public static final DateTimeFormatter PARSER_FORMATTER = new DateTimeFormatterBuilder()
+//          .appendLiteral("T")
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/FHIRPathTimeValue.java:31]: can we remove it?

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java [hunk 3]
```diff
@@ -1068,19 +1079,20 @@ private IllegalArgumentException unexpectedNumberOfArguments(int arity, String f
...
             debug(ctx);
-            return singleton(FHIRPathDateTimeValue.dateTimeValue(ctx.getText().substring(1)));
+            return singleton(dateTimeValue(ctx.getText().substring(1)));
         }
 
         @Override
         public Collection<FHIRPathNode> visitTimeLiteral(FHIRPathParser.TimeLiteralContext ctx) {
             debug(ctx);
-            return singleton(FHIRPathTimeValue.timeValue(ctx.getText().substring(1)));
+//          return singleton(timeValue(ctx.getText().substring(1)));
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1094]: can we remove it?

File: fhir-path/src/main/java/com/ibm/fhir/path/FHIRPathDateTimeValue.java (hunk 1 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/FHIRPathDateTimeValue.java:124]: Removed

File: fhir-path/src/main/java/com/ibm/fhir/path/FHIRPathTimeValue.java (hunk 2 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/FHIRPathTimeValue.java:31]: Removed

File: fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java (hunk 3 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/evaluator/FHIRPathEvaluator.java:1094]: Removed

File: fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java [hunk 4]
```diff
@@ -415,6 +588,29 @@ public static TemporalAmount getTemporalAmount(FHIRPathQuantityValue quantityVal
...
+        case "minutes":
+            return ChronoUnit.MINUTES;
+        case "seconds":
+            return ChronoUnit.SECONDS;
+        case "milliseconds":
+            return ChronoUnit.MILLIS;
+        default:
+            throw new IllegalArgumentException("Unsupported unit: " + unit);
+        }
+    }
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java:612]: should it support variations of those string units per the fhirpath spec

File: fhir-path/src/main/java/com/ibm/fhir/path/function/BetweenFunction.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import java.math.BigDecimal;
+import java.time.temporal.ChronoUnit;
+import java.time.temporal.Temporal;
+import java.util.Collection;
+import java.util.List;
+
+import com.ibm.fhir.path.FHIRPathNode;
+import com.ibm.fhir.path.evaluator.FHIRPathEvaluator.EvaluationContext;
+
+public class BetweenFunction extends FHIRPathAbstractFunction {
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/BetweenFunction.java:27]: javadoc?

File: fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java [hunk 6]
```diff
@@ -270,41 +302,166 @@ public static boolean hasQuantityValue(Collection<FHIRPathNode> nodes) {
...
+     *    true if the collection of nodes evaluates to a boolean, false otherwise
+     */
+    public static boolean evaluatesToTrue(Collection<FHIRPathNode> nodes) {
+        if (convertsToBoolean(nodes)) {
+            return toBoolean(nodes) == true;
+        }
+        return isSingleton(nodes);
     }
 
     public static boolean isTrue(Collection<FHIRPathNode> nodes) {
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java:324]: could use some javadoc to further distinguish it from evaluatesToTrue;  this one would basically just say it returns true iff it convertsToBoolean and has a true value

File: fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java [hunk 7]
```diff
@@ -270,41 +302,166 @@ public static boolean hasQuantityValue(Collection<FHIRPathNode> nodes) {
...
+        return convertsToBoolean(nodes) && toBoolean(nodes) == true;
+    }
+
+    public static boolean isFalse(Collection<FHIRPathNode> nodes) {
+        return convertsToBoolean(nodes) && toBoolean(nodes) == false;
+    }
+
+    public static boolean convertsToBoolean(Collection<FHIRPathNode> nodes) {
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java:332]: maybe some javadoc here with a link to explain that returns true if the value can be implicitly or explicitly converted to a boolean as of that table in the spec

File: fhir-path/src/test/java/com/ibm/fhir/path/test/BetweenFunctionTest.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    public void testBetweenFunction11() throws Exception {
+        FHIRPathEvaluator evaluator = FHIRPathEvaluator.evaluator();
+        try {
+            evaluator.evaluate("between(@1970-01-01T00:00:00Z, @2020-01-01, 'years')");
+            fail();
+        } catch (Exception e) {
+            assertTrue(e instanceof FHIRPathException);
+            assertTrue(e.getCause() instanceof DateTimeException);
+        }
+    }
```
Comment (lmsurpre) [inline fhir-path/src/test/java/com/ibm/fhir/path/test/BetweenFunctionTest.java:108]: could use TestNG expectedExceptions, but this is fine

File: fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java (hunk 4 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java:637]: I wanted to simplify it and only allow the equivalent of what the underlying java.time library allows

File: fhir-path/src/main/java/com/ibm/fhir/path/function/BetweenFunction.java (hunk 5 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/BetweenFunction.java:31]: Added Javadoc

File: fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java (hunk 6 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java:332]: Added Javadoc

File: fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java (hunk 7 above)
Comment (JohnTimm) [reply fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java:357]: Added Javadoc

File: fhir-path/src/test/java/com/ibm/fhir/path/test/BetweenFunctionTest.java (hunk 8 above)
Comment (JohnTimm) [reply fhir-path/src/test/java/com/ibm/fhir/path/test/BetweenFunctionTest.java:108]: I prefer this approach because I can check exception type and message

**Pull request merged**